src/shared/metrics.c

```diff
@@ -0,0 +1,222 @@
+
+        r = sd_varlink_server_bind_method_many(
+                s,
+                "io.systemd.Metrics.List", vl_method_list_cb,
+                "io.systemd.Metrics.Describe", vl_method_describe_cb);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to register varlink metrics methods: %m");
+
+        r = sd_varlink_server_attach_event(s, event, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink metrics connection to event loop: %m");
+
+        *server = TAKE_PTR(s);
+
+        return 0;
+}
+
+int metrics_listen_varlink_address(sd_varlink_server *server, const char *address) {
+        int r;
+
+        assert(server);
+        assert(address);
+
+        /* a new server will have empty list of addresses anyway */
+        if (varlink_server_contains_socket(server, address))
+                return 0;
+
+        r = sd_varlink_server_listen_address(server, address, 0666 | SD_VARLINK_SERVER_MODE_MKDIR_0755);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind to metrics varlink socket '%s': %m", address);
+
+        return 0;
+}
+
+static const char * const metric_family_type_table[_METRIC_FAMILY_TYPE_MAX] = {
+        [METRIC_FAMILY_TYPE_COUNTER] = "counter",
+        [METRIC_FAMILY_TYPE_GAUGE]   = "gauge",
+        [METRIC_FAMILY_TYPE_STRING]  = "string",
+};
+
+DEFINE_STRING_TABLE_LOOKUP_TO_STRING(metric_family_type, MetricFamilyType);
+
+static int metric_family_build_json_one(sd_varlink *link, const MetricFamily* mf, bool more) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(mf);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("name", mf->name),
+                        SD_JSON_BUILD_PAIR_STRING("description", mf->description),
+                        SD_JSON_BUILD_PAIR_STRING("type", metric_family_type_to_string(mf->type)));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+int metrics_method_describe(const MetricFamily metric_family_table[], sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        const MetricFamily *previous = NULL;
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
```

Comment:
I would combine everything in one loop:
```c
for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
  bool more =  ((mf + 1) && (mf+1)->name);
 r = metric_family_build_json_one(link, mf, more);
 if (r < 0)
        return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
}
```

Reply:
same here. Previous is a common pattern so I'd keep it unless a stronger reason